### ISSUE TITLE
Future of morepath

### ISSUE THREAD

**Issue by yeraydiazdiaz** [morepath/morepath#535]
Hi all, while browsing the new PyPI I noticed the last release 0.18.1 was almost a year ago and recent commits are centered on documentation and never quite made it to the live docs.

I was wondering what are the morepath community's thoughts in regards to its future.

Please don't take this the wrong way, I truly feel morepath is an excellent project that has a lot to offer and saddens me to see it die a slow death.

My hope is that this issue will spark a healthy discussion around the project's direction and future.

**Comment by href** [comment]
Hi, thanks for bringing this up!

At my company we are quite invested into Morepath's future. We are going to carry our Morepath based stack around for many years to come and as such will always make sure that it at least works as well as it does now. We do not run a fork of Morepath so if anything breaks we have no choice but to fix it and push out a new release.

Now I'd be happy to see new features, better documentation and further development for Morepath, but I personally do not have any extra time to invest. Morepath is rock solid for us and as such doesn't need much of our attention :)

What I sometimes try to do is create a `more.*` package instead of writing something more integrated into our own framework, but at this point that's not something that occurs often. As time goes on our needs tend to be very specific and not easily generalisable.

That's my 50 cents. I'd love to hear from everyone else what they are using Morepath for and where they see themselves in the community. I guess our place at this point is to be a party that can ensure that maintenance releases are published when needed.

**Comment by faassen** [comment]
Hey, this is a good question! 

I have been driving Morepath for the first few years, but unfortunately I've been able to spend much less time on Morepath in the last year or more, because of a customer that had already made another backend choice. This stays this way for the foreseeable future. Perhaps next year things will change, but it's impossible to predict. Luckily Morepath is quite a compact project and it works well, so it doesn't need a lot of maintenance (there's a Python 3 issue for Reg that was reported that I hope someone can look into though!).

I'd love to be able to build something with Morepath. Some vague plans:

* we need a proper Morepath website.

* I have had a plan to create a "most" project which is more an all-in framework ala Django that deliberately makes technology choices (ORM, etc) so that people don't have to make those choices. But so little time!

* asynchronous Morepath. What if resolving the path to a model were an asynchronous request? Maybe it won't really mean much gain, but it's interesting to speculate.

What are your ideas about the project's direction and future?

**Comment by yeraydiazdiaz** [comment]
I'm really glad to hear both of your commitments to the project.

Personally I haven't had the chance to put Morepath in production, mainly because my clients already have their decision in regards to a web framework, so my experience with it is somewhat limited. However it really resonates with my mental model I'd love to make a case for it in the future.

Given you both mention the stability and low maintenance of Morepath, I'd like to see a push for a 1.0 release and publicise it starting, of course, with a proper website as @faassen mentions. Possibly including:

- case studies, maybe based on @href's experience as a starting point,
- benchmarks, which I believe Morepath did quite well recently
- a tutorial, [I've started one following Miguel Grinberg's Flask Mega-Tutorial](https://github.com/yeraydiazdiaz/morepath_microblog/)
- list of `more.*` packages and integrations

I'm less clear on what features should be included or not in the 1.0 release. @href might have a clearer view of what he's missed in production, and if they'd be part of core or a `more.*` integration. Maybe some integrations should considered "official" or "core" similar to what [Vue.js](https://vuejs.org/) does which I think works well, maybe SQLAlchemy and Jinja which are the most popular ones?

**Comment by faassen** [comment]
It's harder to get a new web framework accepted by a client because there are so many big fish and people basically think there is no real difference or improvement possible.

I'd be happy to help out with a website and even write quite a bit of text, but I need someone else (you?) to take the lead on it. We also need a nice design; @blaflamme in the past offered help with that.

**Comment by henri-hulski** [comment]
I'm using Morepath as a backend atm on some small projects, but planning something bigger.
It's the backend framework of my choice.
I've also setup a [Morepath realworld example app](https://github.com/yacoma/morepath-realworld-example-app) inspired by the https://realworld.io project.
Just missing here the CORS integration and really need someone to review https://github.com/morepath/more.cors/pull/2 so that I can release more.cors.
It's really hard to go on when I'm not even able to get a review.
What should I do in such a case? Just merge the PR?

**Comment by href** [comment]
> I'm less clear on what features should be included or not in the 1.0 release. @href might have a clearer view of what he's missed in production, and if they'd be part of core or a more.* integration. Maybe some integrations should considered "official" or "core" similar to what Vue.js does which I think works well, maybe SQLAlchemy and Jinja which are the most popular ones?

Actually I wouldn't say Morepath needs anything for a 1.0 release at this point. I think we could reasonably call it 1.0 and be done with it. As a micro-framework it doesn't have to come with any bells and whistles.

When it comes to integrations that's of course another issue. There are some implementations that we've written for ourselves that others might find useful. Browser sessions and e-mail sending comes to mind.

However it's not altogether simple to write an integration package when you are the only user. Great integration packages in other frameworks are great when they are used by many projects, as it forces them to really abstract all the non-essential stuff away.

So currently we can sometimes identify something that is reasonably limited in its scope to maybe have it work with other applications as well (like my most recent [more.content_security](https://github.com/morepath/more.content_security) package). But often this is not that easy and writing an integration with some vague assumptions about how others might use it is not a good recipe for success.

So ideally we would have a bunch of different browser session implementations used in a morepath application and we would then be able to extract their essence out into an integration that would rule them all. Starting with the integration before you even have the real-world usage seems to me like putting the cart before the horse.

That is to say we would probably need more users that are willing to use Morepath even though it doesn't come with batteries included. That was basically why went with it. The model of Morepath fit our thinking and our way of working. We've created multiple applications that we basically sell as standard software and we've spawned new customised instances of those applications for lucrative projects. We chose Morepath because we've worked like that before when our code base was based on Zope/Plone.

**Comment by yeraydiazdiaz** [comment]
@faassen I don't mind driving the website efforts, I'll hijack #503 if that's ok.

@href it's great to hear you think the code is already stable enough to call it 1.0. Regarding the integrations I agree we need a bigger user base to properly tests them, my point is to have at least one high-quality solution for very common use cases before we start publicising 1.0.

For instance chances are a user will want an ORM, we could define that our "core" ORM integration is with SQLAlchemy (or PonyORM or whichever we decide) and meet that user's expectations for the use case.

Which use cases we'd like include is up for debate and probably depends on how different users make use of Morepath.

**Comment by href** [comment]
I'd say it makes sense to list the integrations somewhere and maybe have endorsements/upvotes for them. Every time someone says "yeah I use that successfully in production" we give it a bit more weight.

**Comment by henri-hulski** [comment]
@faassen
> I just commented and left a tiny code style issue there

Hmm could not find anything.  :)

**Comment by kagesenshi** [comment]
> Separately from this someone could create a separate framework (that I call Webmost) that doesn't just endorse integrations but actively makes choices. If you install webmost, you get those choices. Since it's Morepath underneath you could make different choices, but the goal of Webmost is different from the goal of Morepath. Morepath is a powerful engine with solid support for those who want to make their own choices. Webmost is a "batteries included" web framework that's a one-stop shop and makes choices for you. People don't always want to have to make choices, because they're in a hurry, or because they are new to web development. Those people go to Django or some other integrated framework. (now people also go to Django because it's popular and has a huge community, which are good reasons, but we can't fix that). Anyway, this is all theoretical currently: there is no Webmost and I don't have time to do the work right now.

I was building one over the past year as my team is also now deeply invested in morepath. Finally got around to document it to a certain degree. It's split into 2 parts, morpfw and morpcc. morpfw is the backend REST API framework, while morpcc builds a UI layer on top of morpfw. 

* morpfw docs - http://morpframework.rtfd.org 
* morpcc docs - http://morpcc.rtfd.org

It takes a lot of inspiration from Plone content types because training junior devs to build authentication & non-messy CRUD is hard.
